clients/cobol-lsp-vscode-extension/src/services/ControlFlowService.ts

```diff
@@ -103,75 +127,155 @@ class AnalysisTask {
 
 export class ControlFlowAnalysisService implements AnalysisServiceDelegate {
   private tasks: Map<string, AnalysisTask>;
-  private callbacks: Map<string, ControlFlowAnalysisCallback>;
+  private latestResults: Map<string, LatestResultData>;
   private diagnosticService: DiagnosticService;
+  private requestVersion: number = 0;
 
   public constructor(
     private mainChannel?: vscode.OutputChannel,
     private logChannel?: vscode.LogOutputChannel,
   ) {
     this.tasks = new Map<string, AnalysisTask>();
-    this.callbacks = new Map<string, ControlFlowAnalysisCallback>();
     this.diagnosticService = new DiagnosticService();
+    this.latestResults = new Map<string, LatestResultData>();
+    this.requestVersion = 1;
+  }
+
+  public async invalidate(documentUri: string, rejectPromise: boolean) {
+    this.logChannel?.debug(`Invalidate document: ${documentUri}`);
+
+    const latestResult = this.latestResults.get(documentUri);
+    if (latestResult) {
+      if (rejectPromise) {
+        this.latestResults.delete(documentUri);
+        latestResult.reject("invalidate");
+      } else {
+        latestResult.requestVersion = 0;
```

Comment:
Won't this leave the promise unresolved for ever?

The analysis API is called, you start typing before the analysis is done, the `requestVersion` is set to 0 and therefore it will never match in the finishTask.

Reply:
But if you are typing new api result from backend will come and will be handled by the handler, after that the version will be updated and a new task with the same version will be created

Reply:
`requestVersion === result?.requestVersion` will never match.

Reply:
Why? I can see logs and they match every time you type something

Reply:
I missed this line `latestResult.requestVersion = this.requestVersion;`...